plugins/storage/volume/linstor/src/main/java/org/apache/cloudstack/storage/datastore/lifecycle/LinstorPrimaryDataStoreLifeCycleImpl.java

```diff
@@ -0,0 +1,327 @@
+    }
+
+    protected boolean createStoragePool(long hostId, StoragePool pool) {
+        s_logger.debug("creating pool " + pool.getName() + " on  host " + hostId);
+
+        if (pool.getPoolType() != Storage.StoragePoolType.Linstor) {
+            s_logger.warn(" Doesn't support storage pool type " + pool.getPoolType());
+            return false;
+        }
+        CreateStoragePoolCommand cmd = new CreateStoragePoolCommand(true, pool);
+        final Answer answer = _agentMgr.easySend(hostId, cmd);
+        if (answer != null && answer.getResult()) {
+            return true;
+        } else {
+            _primaryDataStoreDao.expunge(pool.getId());
+            String msg = answer != null ?
+                "Can not create storage pool through host " + hostId + " due to " + answer.getDetails() :
+                "Can not create storage pool through host " + hostId + " due to CreateStoragePoolCommand returns null";
+            s_logger.warn(msg);
+            throw new CloudRuntimeException(msg);
+        }
+    }
+
+    @Override
+    public boolean attachCluster(DataStore dataStore, ClusterScope scope) {
+        PrimaryDataStoreInfo primaryDataStoreInfo = (PrimaryDataStoreInfo) dataStore;
+
+        // check if there is at least one host up in this cluster
+        List<HostVO> allHosts = resourceMgr.listAllUpAndEnabledHosts(Host.Type.Routing,
+            primaryDataStoreInfo.getClusterId(), primaryDataStoreInfo.getPodId(),
+            primaryDataStoreInfo.getDataCenterId());
+
+        if (allHosts.isEmpty()) {
+            _primaryDataStoreDao.expunge(primaryDataStoreInfo.getId());
+
+            throw new CloudRuntimeException(
+                "No host up to associate a storage pool with in cluster " + primaryDataStoreInfo.getClusterId());
+        }
+
+        List<HostVO> poolHosts = new ArrayList<>();
+        for (HostVO host : allHosts) {
+            try {
+                createStoragePool(host.getId(), primaryDataStoreInfo);
+
+                _storageMgr.connectHostToSharedPool(host.getId(), primaryDataStoreInfo.getId());
+
+                poolHosts.add(host);
+            } catch (Exception e) {
+                s_logger.warn("Unable to establish a connection between " + host + " and " + primaryDataStoreInfo, e);
+            }
+        }
+
+        if (poolHosts.isEmpty()) {
+            s_logger.warn("No host can access storage pool '" + primaryDataStoreInfo + "' on cluster '"
+                + primaryDataStoreInfo.getClusterId() + "'.");
+
+            _primaryDataStoreDao.expunge(primaryDataStoreInfo.getId());
+
+            throw new CloudRuntimeException("Failed to access storage pool");
+        }
+
+        dataStoreHelper.attachCluster(dataStore);
+
+        return true;
+    }
+
+    @Override
+    public boolean attachZone(DataStore dataStore, ZoneScope scope, HypervisorType hypervisorType) {
+        List<HostVO> hosts = resourceMgr.listAllUpAndEnabledHostsInOneZoneByHypervisor(HypervisorType.KVM,
```

Comment:
Hypervisor hard-coded to KVM here, check with the _hypervisorType_ input and if it is supported or not.